source/index.ts

```diff
@@ -0,0 +1,152 @@
+import type {Options, ParsedOptions, Result, AnyFlags} from './types.js';
+
+const buildResult = <Flags extends AnyFlags = AnyFlags>({pkg: packageJson, ...options}: ParsedOptions, parserOptions: ParserOptions): Result<Flags> => {
+	const argv = parseArguments(options.argv as string[], parserOptions);
+	let help = '';
+
+	if (options.help) {
+		help = trimNewlines((options.help || '').replace(/\t+\n*$/, ''));
+
+		if (help.includes('\n')) {
+			help = redent(help, 2);
+		}
+
+		help = `\n${help}`;
+	}
+
+	if (options.description !== false) {
+		let {description} = options;
+
+		if (description) {
+			description = help ? `\n  ${description}\n` : `\n${description}`;
+			help = `${description}${help}`;
+		}
+	}
+
+	help += '\n';
+
+	const showHelp = (code?: number) => {
+		console.log(help);
+		process.exit(typeof code === 'number' ? code : 2);
+	};
+
+	const showVersion = () => {
+		console.log(options.version);
+		process.exit(0);
+	};
+
+	if (argv._.length === 0 && options.argv.length === 1) {
+		if (argv['version'] === true && options.autoVersion) {
+			showVersion();
+		} else if (argv['help'] === true && options.autoHelp) {
+			showHelp(0); // TODO: why is this 0 and showHelp is 2?
+		}
+	}
+
+	const input = argv._ as string[];
+	// TODO:
+	// @ts-expect-error: TS errors bc argv._ does exist, but it thinks we might want to use the type later
+	delete argv._;
+
+	if (!options.allowUnknownFlags) {
+		checkUnknownFlags(input);
+	}
+
+	const flags = camelCaseKeys(argv, {exclude: ['--', /^\w$/]});
+	const unnormalizedFlags = {...flags};
+
+	validate(flags, options);
+
+	for (const flagValue of Object.values(options.flags)) {
+		if (Array.isArray(flagValue.aliases)) {
+			for (const alias of flagValue.aliases) {
+				delete flags[alias];
+			}
+		}
+
+		if (flagValue.shortFlag) {
+			delete flags[flagValue.shortFlag];
+		}
+	}
+
+	checkMissingRequiredFlags(options.flags, flags, input);
+
+	return {
+		input,
+		flags,
+		unnormalizedFlags,
+		pkg: packageJson,
+		help,
+		showHelp,
+		showVersion,
+	} as unknown as Result<Flags>; // TODO: flags complain bc Result camelcases
+};
+
+/**
+@param helpMessage - Shortcut for the `help` option.
+
+@example
+```
+#!/usr/bin/env node
+import meow from 'meow';
+import foo from './index.js';
+
+const cli = meow(`
+	Usage
+	  $ foo <input>
+
+	Options
+	  --rainbow, -r  Include a rainbow
+
+	Examples
+	  $ foo unicorns --rainbow
+	  🌈 unicorns 🌈
+`, {
+	importMeta: import.meta,
+	flags: {
+		rainbow: {
+			type: 'boolean',
+			shortFlag: 'r'
+		}
+	}
+});
+
+//{
+//	input: ['unicorns'],
+//	flags: {rainbow: true},
+//	...
+//}
+
+foo(cli.input.at(0), cli.flags);
+```
+*/
+export default function meow<Flags extends AnyFlags>(helpMessage: string, options?: Options<Flags>): Result<Flags>;
+export default function meow<Flags extends AnyFlags>(options?: Options<Flags>): Result<Flags>;
+// TODO: should these be optional? importMeta is always needed
```

Comment:
Does it make sense for the `options` parameter to be optional? `meow` will just throw anyway without it, since `options.importMeta` is undefined in that case

Reply:
It should not be optional.